pkg/controller/daemon/update.go

```diff
@@ -64,20 +85,275 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+			continue
+		}
+		result = append(result, history)
+	}
+	return result, nil
+}
+
+// match check if ds template is semantically equal to the template stored in history
+func match(ds *extensions.DaemonSet, history *apps.ControllerRevision) (bool, error) {
+	template, err := decodeHistory(history)
+	return apiequality.Semantic.DeepEqual(&ds.Spec.Template, template), err
+}
+
+func decodeHistory(history *apps.ControllerRevision) (*v1.PodTemplateSpec, error) {
+	raw := history.Data.Raw
+	decoder := json.NewDecoder(bytes.NewBuffer(raw))
+	template := v1.PodTemplateSpec{}
+	err := decoder.Decode(&template)
+	return &template, err
+}
+
+func encodeTemplate(template *v1.PodTemplateSpec) ([]byte, error) {
+	buffer := new(bytes.Buffer)
+	encoder := json.NewEncoder(buffer)
+	err := encoder.Encode(template)
+	return buffer.Bytes(), err
+}
+
+// computeHash returns a hash value calculated from pod template and a uniquifier to avoid hash collision
+// TODO: this is copied from deployment controller util, will move to a shared library for all controllers
+func computeHash(template *v1.PodTemplateSpec, uniquifier *int64) uint32 {
+	podTemplateSpecHasher := fnv.New32a()
+	hashutil.DeepHashObject(podTemplateSpecHasher, *template)
+
+	// Add uniquifier in the hash if it exists.
+	if uniquifier != nil {
+		uniquifierBytes := make([]byte, 8)
+		binary.LittleEndian.PutUint64(uniquifierBytes, uint64(*uniquifier))
+		podTemplateSpecHasher.Write(uniquifierBytes)
+	}
+
+	return podTemplateSpecHasher.Sum32()
+}
+
+func (dsc *DaemonSetsController) snapshot(ds *extensions.DaemonSet, revision int64) (*apps.ControllerRevision, error) {
+	encodedTemplate, err := encodeTemplate(&ds.Spec.Template)
+	if err != nil {
+		return nil, err
+	}
+	hash := fmt.Sprint(computeHash(&ds.Spec.Template, ds.Status.Uniquifier))
+	name := ds.Name + "-" + hash
+	history := &apps.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            name,
+			Namespace:       ds.Namespace,
+			Labels:          labelsutil.CloneAndAddLabel(ds.Labels, extensions.DefaultDaemonSetUniqueLabelKey, hash),
+			Annotations:     ds.Annotations,
+			OwnerReferences: []metav1.OwnerReference{*newControllerRef(ds)},
+		},
+		Data:     runtime.RawExtension{Raw: encodedTemplate},
+		Revision: revision,
+	}
+
+	history, err = dsc.kubeClient.AppsV1beta1().ControllerRevisions(ds.Namespace).Create(history)
+	if errors.IsAlreadyExists(err) {
+		// TODO: Is it okay to get from historyLister?
+		existedHistory, getErr := dsc.kubeClient.AppsV1beta1().ControllerRevisions(ds.Namespace).Get(name, metav1.GetOptions{})
+		if getErr != nil {
+			return nil, getErr
+		}
+		// Check if we already created it
+		done, err := match(ds, existedHistory)
+		if err != nil {
+			return nil, err
+		}
+		if done {
+			return existedHistory, nil
+		}
+
+		// Handle name collisions between different history
+		// TODO: Is it okay to get from dsLister?
```

Comment:
I feel it's safer to get from API server, and we only do this call upon name collisions so the performance impact will be low